pkg/policy/defaultRules_test.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/ghodss/yaml"
+	"github.com/stretchr/testify/assert"
+)
+
+//go:embed defaultRulesSchema.json
+var defaultRulesSchemaContent string
+
+const defaultRulesYamlPath = "./defaultRules.yaml"
+
+func TestDefaultRulesFileExists(t *testing.T) {
+	_, fileReadError := getFileFromPath(defaultRulesYamlPath)
+
+	assert.Nil(t, fileReadError)
+}
+
+func TestDefaultRulesFileFitsJSONSchema(t *testing.T) {
+	validationError := validateYamlUsingJSONSchema(defaultRulesYamlPath, defaultRulesSchemaContent)
+
+	assert.Nil(t, validationError)
+}
+
+func TestDefaultRulesHasUniqueNamesInRules(t *testing.T) {
+	defaultRulesFileContent, _ := getFileFromPath(defaultRulesYamlPath)
+
+	defaultRulesFileContentJSON, conversionToJSONError := convertYamlFileContentToJSON(defaultRulesFileContent)
+	assert.Nil(t, conversionToJSONError)
+
+	uniquenessValidationError := validateUniqueStringPropertyValuesInArray("uniqueName", defaultRulesFileContentJSON["rules"])
+	assert.Nil(t, uniquenessValidationError)
+}
+
+func TestDefaultRulesHasUniqueIDsInRules(t *testing.T) {
+	defaultRulesFileContent, _ := getFileFromPath(defaultRulesYamlPath)
+
+	defaultRulesFileContentJSON, conversionToJSONError := convertYamlFileContentToJSON(defaultRulesFileContent)
+	assert.Nil(t, conversionToJSONError)
+
+	uniquenessValidationError := validateUniqueFloat64PropertyValuesInArray("id", defaultRulesFileContentJSON["rules"])
+
+	assert.Nil(t, uniquenessValidationError)
+}
+
+func getFileFromPath(path string) (string, error) {
+	fileReader := fileReader.CreateFileReader(nil)
+	fileContent, readFileError := fileReader.ReadFileContent(path)
+
+	if readFileError != nil {
+		return "", readFileError
+	}
+
+	return fileContent, nil
+}
+
+func validateYamlUsingJSONSchema(yamlFilePath string, schema string) error {
+	fileContent, _ := getFileFromPath(yamlFilePath)
+	jsonSchemaValidator := jsonSchemaValidator.New()
+	schemaValidationResult, schemaValidationError := jsonSchemaValidator.ValidateYamlSchema(schema, fileContent)
+
+	if schemaValidationError != nil {
+		return schemaValidationError
+	}
+
+	if !schemaValidationResult.Valid() {
+		validationErrors := fmt.Errorf("Received validation errors for %s:\n", yamlFilePath)
+
+		for _, validationError := range schemaValidationResult.Errors() {
+			validationErrors = fmt.Errorf("%s\n%s", validationErrors, validationError)
+		}
+
+		return validationErrors
+	}
+
+	return nil
+}
+
+func convertYamlFileContentToJSON(yamlFileContent string) (map[string][]interface{}, error) {
+	yamlFileContentRawJSON, yamlParseError := yaml.YAMLToJSON([]byte(yamlFileContent))
+
+	if yamlParseError != nil {
+		return map[string][]interface{}{}, yamlParseError
+	}
+
+	var yamlFileContentJSON map[string][]interface{}
+	jsonMarshallingError := json.Unmarshal(yamlFileContentRawJSON, &yamlFileContentJSON)
+
+	var jsonMarshallingFailed = jsonMarshallingError != nil && reflect.TypeOf(yamlFileContentJSON) != reflect.TypeOf(map[string][]interface{}{})
+	if jsonMarshallingFailed {
```

Comment:
The JSON unmarshalling threw the following error,
```
 nil, but got: &json.UnmarshalTypeError{Value:"strin
g", Type:(*reflect.rtype)(0x13f3b00), Offset:139, Struct:"", Field:""}
```
I added the extra condition as a hack to prevent golang from panicing when the unmarshalling does a good job.
The tests worked with that code so I let it be when I worked on it,
definitely not proud of it in any way.

Would love to get some help with solving it in a different way 🙏

Reply:
Ok I think I was able to understand what is the best here.
First of all, the `aliases` in `defaultRules.yaml` is a but problematic when reading the file and converting it to json. We use an alias there and for some reason it makes problems.

I suggest the following, tell me what do you think:
- create a struct that represents the `deafultRules.yaml`:

```type DefaultRules struct {
	ApiVersion string                  `json:"apiVersion"`
	Rules      []DefaultRuleDefinition `json:"rules"`
}
```

- When you Unmarshal you can do like this:
```
	var yamlFileContentJSON DefaultRules
	jsonUnmarshallingError := json.Unmarshal(yamlFileContentRawJSON, &yamlFileContentJSON)
```

And instead of using `defaultRulesFileContentJSON["rules"]` you can pass `defaultRulesFileContentJSON.Rules`.
In each place use the new type `DefaultRules`.

Example for a refactored function:
```
for _, item := range defaultRules {
		if propertyValuesExistenceMap[item.UniqueName] {
			return fmt.Errorf("property %s has duplicate value %s", propertyName, item.UniqueName)
		}

		propertyValuesExistenceMap[item.UniqueName] = true
	}
```

I know you tried to make the function as generic as they can be but I think that in this case it's ok to make them especially for the validation of the file.

This way we don't need the reflect code and the code that Unmarshal to an unknown map.

Let me know if it was clear enough